transaction/tests/DeadLockDetector_unittest.cpp

```diff
@@ -0,0 +1,76 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "transaction/DeadLockDetector.hpp"
+
+#include <atomic>
+#include <memory>
+#include <vector>
+
+#include "transaction/DirectedGraph.hpp"
+#include "transaction/LockTable.hpp"
+#include "transaction/Transaction.hpp"
+
+#include "gtest/gtest.h"
+
+namespace quickstep {
+namespace transaction {
+
+class DeadLockDetectorTest : public ::testing::Test {
+ protected:
+  virtual void SetUp() {
+    lock_table_ = std::make_unique<LockTable>();
+    status_.store(DeadLockDetectorStatus::kDone);
+  }
+
+  std::unique_ptr<LockTable> lock_table_;
+  std::atomic<DeadLockDetectorStatus> status_;
+  std::vector<DirectedGraph::node_id> victims_;
+};
+
+TEST_F(DeadLockDetectorTest, SimpleCycle) {
+  lock_table_->putLock(transaction_id(1),
+                       ResourceId(4, 5),
+                       AccessMode(AccessModeType::kXLock));
+
+  lock_table_->putLock(transaction_id(2),
+                       ResourceId(1, 2),
+                       AccessMode(AccessModeType::kXLock));
+
+  lock_table_->putLock(transaction_id(1),
+                       ResourceId(1, 2),
+                       AccessMode(AccessModeType::kXLock));
+
+  lock_table_->putLock(transaction_id(2),
+                       ResourceId(4, 5),
+                       AccessMode(AccessModeType::kXLock));
```

Comment:
Well, I think we could refactor the code more clearly.

```
  const transaction_id transaction_one(1), transaction_two(2).
  const ResourceId resource_one(...), resource_two(...);
  const AccessMode x_lock(AccessModeType::kXLock);
  
  // Create an interleaving deadlock among two transactions and two resources
  // where each transaction holds one resource, and tries to acquire the other resource.
  lock_table_->putLock(transaction_one, resource_one, x_lock);
  lock_table_->putLock(transaction_two, resource_two, x_lock);
  
  lock_table_->putLock(transaction_one, resource_two, x_lock);
  lock_table_->putLock(transaction_two, resource_one, x_lock);
```

The code might not compile, but you should get the sense that we should explicitly express the interleaving deadlock by using variable names and how to add comments.